test/stateful_dataloader/test_state_dict.py

```diff
@@ -1540,26 +1542,52 @@ def _get_iter_calls(self, state):
         else:
             w_states = list(state["_snapshot"]["_worker_snapshots"].values())
 
-        return [x["dataset_state"]["iter_calls"] for x in w_states]
+        if w_states[0]["dataset_state"] is not None:
+            return [x["dataset_state"]["iter_calls"] for x in w_states]
+        return [x["fetcher_state"]["dataset_iter_state"]["iter_calls"] for x in w_states]
 
     def _run_test(self, num_workers, dataset):
+        # Need to make a copy here as iter calls is tracked and its count is stored in dataset state which persists across calls for single process runs.
+        dataset_copy = deepcopy(dataset) if num_workers == 0 else dataset
         dl = StatefulDataLoader(
             dataset=dataset,
             num_workers=num_workers,
             multiprocessing_context=("forkserver" if IS_MACOS and num_workers else None),
+            persistent_workers=True if num_workers else False,
         )
-        iter(dl)
+        it = iter(dl)
+        state = dl.state_dict()
+        # Ensure iter is called only once per worker
+        self.assertEqual(self._get_iter_calls(state), [1] * max(1, num_workers))
+
+        for _ in range(10):
+            next(it)
         state = dl.state_dict()
+        # Ensure that iter has not been invoked again
         self.assertEqual(self._get_iter_calls(state), [1] * max(1, num_workers))
+
+        # Call iter on dl to see if iter is called again
+        iter(dl)
+        state = dl.state_dict()
+        self.assertEqual(self._get_iter_calls(state), [2] * max(1, num_workers))
+
         dl2 = StatefulDataLoader(
-            dataset=dataset,
+            dataset=dataset_copy,
             num_workers=num_workers,
             multiprocessing_context=("forkserver" if IS_MACOS and num_workers else None),
+            persistent_workers=True if num_workers else False,
         )
         dl2.load_state_dict(state)
-        iter(dl2)
+        it = iter(dl2)
         state2 = dl2.state_dict()
-        self.assertEqual(self._get_iter_calls(state2), [2] * max(1, num_workers))
+        # Ensure that iter is called only once per worker even when dataloader resumes from a state
+        self.assertEqual(self._get_iter_calls(state2), [3] * max(1, num_workers))
+
+        for _ in range(10):
+            next(it)
+        state = dl2.state_dict()
+        # Ensure that iter has not been invoked again
+        self.assertEqual(self._get_iter_calls(state2), [3] * max(1, num_workers))
```

Comment:
Let's change this test to just assert that __iter__ is only called once for new dataloaders with and without state-resumes, but not enforce the actual value of state["iter_calls"]